### ISSUE TITLE
feat: Return application level errors as a part of the gRPC response

### ISSUE THREAD

**Issue by erezrokah** [cloudquery/plugin-sdk#462]
TLDR: We should consider having application level errors as a part of the gRPC responses, to make it easier to identify them.

For example in the following line:
https://github.com/cloudquery/plugin-sdk/blob/ab7ca972e0b187a7dfb66132a03f07479cd29bb7/clients/destination.go#L303
`failed to CloseAndRecv` is not necessarily correct, as `err` is whatever is returned from the following function:
https://github.com/cloudquery/plugin-sdk/blob/ab7ca972e0b187a7dfb66132a03f07479cd29bb7/internal/servers/destinations.go#L65.

Not sure what's the best solution (need to do some research), but we could add an error(s) field to gRPC responses so we have a clear separation between application level errors and gRPC/communication level ones.

Also due to https://github.com/grpc/grpc-go/issues/2934 we can't unwrap errors.

**Comment by erezrokah** [comment]
Looks like we could use `WithDetails` for this https://pkg.go.dev/google.golang.org/grpc/internal/status#Status.WithDetails

**Comment by yevgenypats** [comment]
I understand the need for application level errors but I don't think it should have anything with grpc errors. depending on which errors we want to send we need to add it to the protobuf. grpc errors are mostly used when something bad happened on the server.

**Comment by erezrokah** [comment]
> I understand the need for application level errors

Most errors returned from https://github.com/cloudquery/plugin-sdk/blob/62692b9cb8a3ff3465d9d14a1ec7cc801d3490af/internal/servers/destinations.go#L65 are application level.

The only ones that could be related to gRPC protocol are:
https://github.com/cloudquery/plugin-sdk/blob/62692b9cb8a3ff3465d9d14a1ec7cc801d3490af/internal/servers/destinations.go#L73
https://github.com/cloudquery/plugin-sdk/blob/62692b9cb8a3ff3465d9d14a1ec7cc801d3490af/internal/servers/destinations.go#L101, but also not necessarily depending on what error caused the stream to close

**Comment by yevgenypats** [comment]
There can be also grpc errror returned after this function is returned that we actually saw when buffer is too big so there is no good way to distinguish both over the grpc error field.

**Comment by erezrokah** [comment]
> no good way to distinguish both over the grpc error field

Agree there could be both application errors and gRPC errors. Right now we're mixing both so the purpose of this issue is to discuss a solution so we can distinguish between them.

One approach is in https://github.com/cloudquery/plugin-sdk/pull/487, to use the `status` package as it essentially wraps the errors that are sent, along with the `errdetails` package. More in https://cloud.google.com/apis/design/errors#error_details

**Comment by erezrokah** [comment]
Closing as won't fix as we've improved error reporting
